src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -406,147 +378,140 @@ private static int BinarySearch(T[] array, int index, int length, T value)
         {
-            Debug.Assert(!keys.IsEmpty);
-
-            int lo = 0;
-            int hi = keys.Length - 1;
-
-            // Compute median-of-three.  But also partition them, since we've done the comparison.
-            int middle = lo + ((hi - lo) / 2);
+            Debug.Assert(keys.Length >= Array.IntrosortSizeThreshold);
 
-            // Sort lo, mid and hi appropriately, then pick mid as the pivot.
-            SwapIfGreater(keys, lo, middle);  // swap the low with the mid point
-            SwapIfGreater(keys, lo, hi);   // swap the low with the high
-            SwapIfGreater(keys, middle, hi); // swap the middle with the high
+            // Use median-of-three to select a pivot. Grab a reference to the 0th, Length-1th, and Length/2th elements, and sort them.
+            ref T zeroRef = ref MemoryMarshal.GetReference(keys);
+            ref T lastRef = ref Unsafe.Add(ref zeroRef, keys.Length - 1);
+            ref T middleRef = ref Unsafe.Add(ref zeroRef, (keys.Length - 1) >> 1);
+            SwapIfGreater(ref zeroRef, ref middleRef);
+            SwapIfGreater(ref zeroRef, ref lastRef);
+            SwapIfGreater(ref middleRef, ref lastRef);
 
-            T pivot = keys[middle];
-            Swap(keys, middle, hi - 1);
-            int left = lo, right = hi - 1;  // We already partitioned lo and hi and put the pivot in hi - 1.  And we pre-increment & decrement below.
+            // Select the middle value as the pivot, and move it to be just before the last element.
+            ref T nextToLastRef = ref Unsafe.Add(ref zeroRef, keys.Length - 2);
+            T pivot = middleRef;
+            Swap(ref middleRef, ref nextToLastRef);
 
-            while (left < right)
+            // Walk the left and right pointers, swapping elements as necessary, until they cross.
+            ref T leftRef = ref zeroRef, rightRef = ref nextToLastRef;
+            while (Unsafe.IsAddressLessThan(ref leftRef, ref rightRef))
             {
                 if (pivot == null)
                 {
-                    while (left < (hi - 1) && keys[++left] == null) ;
-                    while (right > lo && keys[--right] != null) ;
+                    while (Unsafe.IsAddressLessThan(ref leftRef, ref nextToLastRef) && (leftRef = ref Unsafe.Add(ref leftRef, 1)) == null) ;
+                    while (Unsafe.IsAddressGreaterThan(ref rightRef, ref zeroRef) && (rightRef = ref Unsafe.Add(ref rightRef, -1)) == null) ;
                 }
                 else
                 {
-                    while (pivot.CompareTo(keys[++left]) > 0) ;
-                    while (pivot.CompareTo(keys[--right]) < 0) ;
+                    while (Unsafe.IsAddressLessThan(ref leftRef, ref nextToLastRef) && pivot.CompareTo(leftRef = ref Unsafe.Add(ref leftRef, 1)) > 0) ;
+                    while (Unsafe.IsAddressGreaterThan(ref rightRef, ref zeroRef) && pivot.CompareTo(rightRef = ref Unsafe.Add(ref rightRef, -1)) < 0) ;
```

Comment:
@stephentoub this will no longer result in an index-out-of-range exception thrown when bogus comparable, insteead, it silently swallows that case now.

Reply:
Yes.  Incorrect comparables wouldn't have always done so, only for some forms of incorrect.

Reply:
Not sure I understand, does that mean you think this is not a problem?

Reply:
Yes.  You disagree?

Reply:
I would have thought that it would be a breaking change, you will no longer get an exception. Behavior is different from earlier versions. Is it not? 

I have no hard feelings about this, clearly my bar for acceptable changes, perf regressions etc. was set too high in some regards 😅 When I worked on this I was aiming for 100% fidelity with existing output.

Reply:
My thinking when I chose to make this change was that the comparer is busted, and there are a multitude of possible behaviors that could result from a busted comparer... it could throw arbitrary exceptions, it could sporadically yield an incorrect comparison in a way no one would know, it could crash the process, etc.  We could even be changing behavior in that regard just by changing how many times we invoked the comparer, or by changing the order in which we invoked it on the input data, or any number of other things.  So I'm not concerned with taking a failure we may have only sometimes been able to detect (and for which a goal was never true detection) and making it into something which sometimes fails differently, and does so by not throwing instead of throwing.

@jkotas, any concerns?

Reply:
I do not have concerns about the subtle behavior change with broken comparers. My primary concerns around the unsafe code are buffer overruns. It is easy to see that the bounds are checked in this case.

> aim for 100% fidelity with existing output.

That was a good place to start and the initial PR that got merged maintained this fidelity.

Reply:
Alright, good to know. Perf/code size will be better without the check :+1: I have tests that will fail due to this change, since I test against built-in `Sort`, but that is my concern then.